### PR TITLE
Add spec test functionality

### PR DESCRIPTION
**Pull request by juliusgeo**

### CODE REVIEW

File: pymongoexplain/commands.py [hunk 1]
```diff
@@ -39,6 +39,8 @@ def command_name(self):
     def get_SON(self):
         cmd = SON([(self.command_name, self.collection)])
         cmd.update(self.command_document)
+        if self.command_document == {}:
+            return {}
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:43]: Why was this change needed?

File: pymongoexplain/commands.py [hunk 2]
```diff
@@ -51,8 +53,13 @@ def __init__(self, collection: Collection, filter, update,
             value = kwargs[key]
             if key == "bypass_document_validation":
                 return_document[key] = value
+            elif key == "hint":
+                return_document["updates"][0]["hint"] = value if \
+                    isinstance(value, str) else SON(value)
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:58]: This does not have the same behavior as Pymongo. Check out how the Collection class handles the "hint" option.

File: pymongoexplain/commands.py [hunk 3]
```diff
@@ -51,8 +53,13 @@ def __init__(self, collection: Collection, filter, update,
...
             else:
                 return_document["updates"][0][key] = value
+        if return_document["updates"][0].get("hint", True) == {}:
+           del return_document["updates"][0]["hint"]
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:62]: Instead of adding and then removing "hint" how about we only add it if needed?

File: pymongoexplain/commands.py [hunk 4]
```diff
@@ -81,7 +88,12 @@ def __init__(self, collection: Collection, pipeline, session,
         super().__init__(collection.name)
         self.command_document = {"pipeline": pipeline, "cursor": cursor_options}
         for key, value in kwargs.items():
-            self.command_document[key] = value
+            if key == "batchSize":
+                if value == 0:
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:92]: Do we need to check for None too?

File: pymongoexplain/commands.py [hunk 5]
```diff
@@ -110,18 +122,30 @@ def __init__(self, collection: Collection,
         super().__init__(collection.name)
         for key, value in kwargs.items():
             self.command_document[key] = value
+        if self.command_document["filter"] == {}:
+            self.command_document = {}
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:126]: Is this correct? What about all the other find command options? Like:
```
coll.find({}, limit=10)
```

File: pymongoexplain/commands.py [hunk 6]
```diff
@@ -110,18 +122,30 @@ def __init__(self, collection: Collection,
...
-            self.command_document[key] = value
+            if key == "hint":
+                self.command_document["hint"] = value if \
+                    isinstance(value, str) else SON(value)
+            else:
+                self.command_document[key] = value
+        if "replacement" in self.command_document.keys():
+            self.command_document["update"] = self.command_document[
+                "replacement"]
+            del self.command_document["replacement"]
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:148]: I think it would be more readable to construct the command correctly from the start rather than constructing an invalid command and then patching it up after.

File: pymongoexplain/explainable_collection.py [hunk 7]
```diff
@@ -16,24 +16,29 @@
...
     UpdateCommand, DistinctCommand, DeleteCommand, FindAndModifyCommand
 
 Document = Union[dict, SON]
 
+
 class ExplainCollection():
-    def __init__(self, collection):
-        self.collection = collection
+    def __init__(self, collection_object):
+        self.collection_object = collection_object
```
Comment (ShaneHarvey) [inline pymongoexplain/explainable_collection.py:30]: I think "collection" is a better name. "collection_object" is a bit verbose for Python.

File: pymongoexplain/commands.py (hunk 1 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:43]: There were a few cases where the filter was {} and the expected command payload in that case was also {}

File: pymongoexplain/commands.py (hunk 2 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:58]: I am confused about this comment. This matches the output produced by Pymongo.

File: pymongoexplain/commands.py (hunk 3 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:62]: Changing it to only be added when needed results in 1 test failing, not sure how to fix it.

File: pymongoexplain/commands.py (hunk 4 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:92]: None values are already removed in the converting to snakecase step

File: pymongoexplain/commands.py (hunk 5 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:126]: This was the only way to make all the tests pass, so I'm not sure.

File: pymongoexplain/explainable_collection.py (hunk 7 above)
Comment (juliusgeo) [reply pymongoexplain/explainable_collection.py:30]: Done.

File: pymongoexplain/commands.py (hunk 6 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:148]: Confused about this comment. The command is being constructed in this step. The kwargs value of replacement has to be patched in that way otherwise it will simply not produce the correct output. If we did not put this code there, then it would have to be duplicated multiple times in explainable_collection.py

Comment (juliusgeo) [reply pymongoexplain/commands.py:148]: There are certain processing steps that must be followed to match pymongo's output, and most of them seem rather arbitrary so the only way for me to verify is to match my code to the test cases.

File: pymongoexplain/commands.py (hunk 2 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:58]: Check out _index_document and where that method is used: https://github.com/mongodb/mongo-python-driver/blob/956ce3d4b0edfa9c1d946109db743f82ed0bfc0a/pymongo/helpers.py#L79

File: pymongoexplain/commands.py (hunk 3 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:62]: Can you post those failures here?

File: pymongoexplain/commands.py (hunk 1 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:43]: It should not be possible for a command to be empty. Could you post those cases?

File: pymongoexplain/commands.py (hunk 5 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:126]: Please add a test for `coll.find({}, limit=10)` to ensure that the limit field (and other command args) are not removed.

File: pymongoexplain/commands.py (hunk 6 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:148]: What I'm trying to get at is that it's confusing to add the "replacement" field to command_document and then immediately remove it. Instead I think this class should handle all the findAndModify specific options while we constructing the command. Like this:
```
    command_document = {
        "findAndModify": collection_name,
        "query": filter,
        "new":  return_document,
        "update": replacement,
    }
    collation = validate_collation_or_none(kwargs.pop('collation', None))
    if collation is not None:
        cmd["collation"] = collation
    if projection is not None:
        cmd["fields"] = helpers._fields_list_to_dict(projection,
                                                     "projection")
    if sort is not None:
        cmd["sort"] = helpers._index_document(sort)
    if upsert is not None:
        common.validate_boolean("upsert", upsert)
        cmd["upsert"] = upsert
    if hint is not None:
        if not isinstance(hint, str):
            hint = helpers._index_document(hint)
        cmd['hint'] = hint
    if array_filters is not None:
        cmd["arrayFilters"] = array_filters
```

This approach exposes numerous bugs in the current approach. Sometimes an argument name needs validation (like a type assertion), sometimes it needs to be transformed (like `collation` or `hint`), and sometimes the python argument name is different from the command argument name (like `replacement` -> `update`).

Comment (juliusgeo) [reply pymongoexplain/commands.py]: I removed the deleting statements and tried to condense all the logic down into the for loops.

File: pymongoexplain/commands.py (hunk 5 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:127]: Added a test.

File: pymongoexplain/commands.py [hunk 8]
```diff
@@ -39,18 +40,25 @@ def command_name(self):
...
-        for key in kwargs:
-            value = kwargs[key]
+        return_document = {
+            "updates":[{"q": filter, "u": update}]
+        }
+        for key, value in kwargs.items():
             if key == "bypass_document_validation":
                 return_document[key] = value
+            elif key == "hint":
+                if value is not {} and value is not None:
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:59]: We don't need the `{}` check here. It's invalid for hint to be `{}`:
```
>>>> client.t.t.update_one({}, {'$set': {'a':1}}, hint={})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/shane/git/mongo-python-driver/pymongo/collection.py", line 1024, in update_one
    hint=hint, session=session),
  File "/Users/shane/git/mongo-python-driver/pymongo/collection.py", line 870, in _update_retryable
    _update, session)
  File "/Users/shane/git/mongo-python-driver/pymongo/mongo_client.py", line 1497, in _retryable_write
    return self._retry_with_session(retryable, func, s, None)
  File "/Users/shane/git/mongo-python-driver/pymongo/mongo_client.py", line 1383, in _retry_with_session
    return self._retry_internal(retryable, func, session, bulk)
  File "/Users/shane/git/mongo-python-driver/pymongo/mongo_client.py", line 1415, in _retry_internal
    return func(session, sock_info, retryable)
  File "/Users/shane/git/mongo-python-driver/pymongo/collection.py", line 866, in _update
    retryable_write=retryable_write)
  File "/Users/shane/git/mongo-python-driver/pymongo/collection.py", line 806, in _update
    hint = helpers._index_document(hint)
  File "/Users/shane/git/mongo-python-driver/pymongo/helpers.py", line 87, in _index_document
    "mean %r?" % list(iteritems(index_list)))
TypeError: passing a dict to sort/create_index/hint is not allowed - use a list of tuples instead. did you mean []?
```

File: pymongoexplain/explainable_collection.py [hunk 9]
```diff
@@ -16,20 +16,25 @@
...
 class ExplainCollection():
     def __init__(self, collection):
         self.collection = collection
         self.last_cmd_payload = None
 
     def _explain_command(self, command):
         command_son = command.get_SON()
+        if command_son == {}:
+            self.last_cmd_payload = {}
+            return {}
```
Comment (ShaneHarvey) [inline pymongoexplain/explainable_collection.py:37]: This should be removed. It's impossible for a command to be empty (`{}`).

File: pymongoexplain/commands.py [hunk 10]
```diff
@@ -108,6 +121,8 @@ class FindCommand(BaseCommand):
     def __init__(self, collection: Collection,
                  kwargs):
         super().__init__(collection.name)
+        if kwargs["filter"] == {}:
+            self.command_document = {}
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:125]: This looks like a bug (or it's masking a bug). Let's remove this code and address the failing test in a different way. Please post the test failure here.

For reference this is how pymongo generates a find command:
https://github.com/mongodb/mongo-python-driver/blob/3.11.0rc0/pymongo/message.py#L181-L217

File: pymongoexplain/commands.py (hunk 1 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:44]: It is in the following cases:
```
SON([('find', 'test_find_allowdiskuse'), ('filter', {})])
{}
```

File: pymongoexplain/commands.py (hunk 8 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:59]: Done.

File: pymongoexplain/commands.py (hunk 10 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:125]: Yeah, I thought that might be a bug because it doesn't really make sense. Removing that doesn't seem to cause any additional test failures.

File: pymongoexplain/explainable_collection.py (hunk 9 above)
Comment (juliusgeo) [reply pymongoexplain/explainable_collection.py:37]: Done.

File: pymongoexplain/commands.py (hunk 1 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:44]: Which spec test is this specifically? Can you post the entire test failure output?

File: pymongoexplain/commands.py (hunk 10 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py]: The three different tests that fail are below, they are in the format of \<the command we constructed\>, \<the command that was expected\>
```
SON([('find', 'test_find_allowdiskuse'), ('filter', {})])
{}
```
```
SON([('find', 'test_find_allowdiskuse'), ('allowDiskUse', True), ('filter', {})])
{}
```
```
SON([('find', 'test_find_allowdiskuse'), ('allowDiskUse', False), ('filter', {})])
{}
```

Comment (ShaneHarvey) [reply pymongoexplain/commands.py]: Can you post the entire failed test output though? Like this:
```
FAIL: test_keyword_arg_defaults (test.test_client.ClientUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/shane/git/mongo-python-driver/test/test_client.py", line 128, in test_keyword_arg_defaults
    self.assertEqual(20.1, pool_opts.connect_timeout)
AssertionError: 20.1 != 20.0
```

Comment (juliusgeo) [reply pymongoexplain/commands.py]:
```
SON([('find', 'test_find_allowdiskuse'), ('filter', {})])
{}

Error
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/Users/julius/pymongoexplain/test/__init__.py", line 442, in wrap
    return f(*args, **kwargs)
  File "/Users/julius/pymongoexplain/test/test_crud_v2.py", line 62, in run_scenario
    self.run_scenario(scenario_def, test)
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 615, in run_scenario
    self.run_test_ops(sessions, collection, test)
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 525, in run_test_ops
    self.run_operations(sessions, collection, test['operations'])
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 431, in run_operations
    result = self.run_operation(sessions, collection, op.copy())
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 374, in run_operation
    self._compare_command_dicts(wrapped_collection.last_cmd_payload,
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 269, in _compare_command_dicts
    self.assertEqual(ours[key], theirs[key])
KeyError: 'find
```

```
SON([('find', 'test_find_allowdiskuse'), ('allowDiskUse', True), ('filter', {})])
{}

Error
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/Users/julius/pymongoexplain/test/__init__.py", line 442, in wrap
    return f(*args, **kwargs)
  File "/Users/julius/pymongoexplain/test/test_crud_v2.py", line 62, in run_scenario
    self.run_scenario(scenario_def, test)
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 615, in run_scenario
    self.run_test_ops(sessions, collection, test)
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 525, in run_test_ops
    self.run_operations(sessions, collection, test['operations'])
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 431, in run_operations
    result = self.run_operation(sessions, collection, op.copy())
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 374, in run_operation
    self._compare_command_dicts(wrapped_collection.last_cmd_payload,
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 269, in _compare_command_dicts
    self.assertEqual(ours[key], theirs[key])
KeyError: 'find'
```
```
SON([('find', 'test_find_allowdiskuse'), ('allowDiskUse', False), ('filter', {})])
{}

Error
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/Cellar/python@3.8/3.8.3_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/Users/julius/pymongoexplain/test/__init__.py", line 442, in wrap
    return f(*args, **kwargs)
  File "/Users/julius/pymongoexplain/test/test_crud_v2.py", line 62, in run_scenario
    self.run_scenario(scenario_def, test)
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 615, in run_scenario
    self.run_test_ops(sessions, collection, test)
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 525, in run_test_ops
    self.run_operations(sessions, collection, test['operations'])
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 431, in run_operations
    result = self.run_operation(sessions, collection, op.copy())
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 374, in run_operation
    self._compare_command_dicts(wrapped_collection.last_cmd_payload,
  File "/Users/julius/pymongoexplain/test/utils_spec_runner.py", line 269, in _compare_command_dicts
    self.assertEqual(ours[key], theirs[key])
KeyError: 'find'

```

File: pymongoexplain/commands.py [hunk 11]
```diff
@@ -44,16 +49,40 @@ def get_SON(self):
...
-        self.command_document = convert_to_camelcase(return_document)
+                 upsert=None, multi=None, collation=None, array_filters=None,
+                 hint=None, ordered=None, write_concern=None,
+                 bypass_document_validation=None, comment=None):
+        super().__init__(collection.name, collation)
+        self.command_document.update({
+            "updates":[{"q": filter, "u": update}]
+        })
+        if upsert is not None:
+            self.command_document["updates"][0]["upsert"] = upsert
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:60]: I suggest refactoring this to:
```
update_doc = {"q": filter, "u": update}
self.command_document["updates"] = [update_doc]
if upsert is not None:
   update_doc["upsert"] = upsert
...
```

```

File: pymongoexplain/commands.py [hunk 12]
```diff
@@ -77,24 +107,33 @@ def command_name(self):
 class AggregateCommand(BaseCommand):
     def __init__(self, collection: Collection, pipeline, session,
                  cursor_options,
-                 kwargs, exclude_keys = []):
-        super().__init__(collection.name)
-        self.command_document = {"pipeline": pipeline, "cursor": cursor_options}
+                 kwargs):
+
+        super().__init__(collection.name, kwargs.get("collation", None))
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:112]: We should be using `kwargs.pop("collation", None)` instead.

File: pymongoexplain/explainable_collection.py [hunk 13]
```diff
@@ -76,7 +75,7 @@ def count_documents(self, filter: Document, session=None,
                                  **kwargs):
 
         command = AggregateCommand(self.collection, [{'$match': filter},
-                                                      {'$group': {'n': {'$sum': 1}, '_id': 1}}],
+                                                     {'$group': {'n': {'$sum': 1}, '_id': 1}}],
                                    session, {}, kwargs,
                                    exclude_keys=filter.keys())
```
Comment (ShaneHarvey) [inline pymongoexplain/explainable_collection.py:80]: What is `exclude_keys`?

File: pymongoexplain/commands.py [hunk 14]
```diff
@@ -63,10 +92,11 @@ def command_name(self):
 class DistinctCommand(BaseCommand):
     def __init__(self, collection: Collection, key, filter, session,
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:93]: Why do some of these classes accept a `session` but some don't? I think they should all be consistent.

File: pymongoexplain/commands.py [hunk 15]
```diff
@@ -63,10 +92,11 @@ def command_name(self):
...
                  kwargs):
-        super().__init__(collection.name)
-        self.command_document = {"key": key, "query": filter}
-        for key, value in kwargs.items():
-            self.command_document[key] = value
+        super().__init__(collection.name, kwargs.get("collation", None))
+        self.command_document.update({"key": key, "query": filter})
+        if kwargs.get("read_concern", None) is not None:
+            self.command_document["read_concern"] = kwargs["read_concern"]
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:98]: `read_concern` is not an option for distinct so we can remove this.

Comment (juliusgeo) [reply pymongoexplain/commands.py:98]: Done.

File: pymongoexplain/commands.py (hunk 12 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:112]: Done.

File: pymongoexplain/commands.py (hunk 11 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:58]: I don't see this change. Did it get lost?

File: pymongoexplain/commands.py [hunk 16]
```diff
@@ -75,26 +101,35 @@ def command_name(self):
...
         return "aggregate"
 
+
 class CountCommand(BaseCommand):
-    def __init__(self, collection: Collection, filter,
-                 kwargs):
-        super().__init__(collection.name)
-        self.command_document = {"query": filter}
+    def __init__(self, collection: Collection, filter, kwargs,
+                 sxclude_keys=None):
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:130]: Remove `sxclude_keys`.

File: pymongoexplain/utils.py [hunk 17]
```diff
@@ -30,12 +30,6 @@ def convert_to_camelcase(d, exclude_keys=[]):
         if "_" in key and key[0] != "_":
             new_key = key.split("_")[0] + ''.join(
                 [i.capitalize() for i in key.split("_")[1:]])
-        if isinstance(d[key], list):
-            ret[new_key] = [convert_to_camelcase(
-                i, exclude_keys=exclude_keys) for i in d[key]]
-        elif isinstance(d[key], dict):
-            ret[new_key] = convert_to_camelcase(d[key],
-                                                     exclude_keys=exclude_keys)
```
Comment (ShaneHarvey) [inline pymongoexplain/utils.py:38]: Great! I think we can remove the `exclude_keys` argument now too.

File: pymongoexplain/explainable_collection.py [hunk 18]
```diff
@@ -149,29 +146,30 @@ def find_one_and_delete(self, filter: Document, projection: list = None,
                                        kwargs)
         return self._explain_command(command)
 
-    def find_one_and_replace(self, filter: Document, replacement: Document,
+    def find_one_and_replace(self, filter: Document, update:
```
Comment (ShaneHarvey) [inline pymongoexplain/explainable_collection.py:149]: This argument should still be called "replacement", not "update". It needs to match Pymongo's find_one_and_replace

File: pymongoexplain/commands.py [hunk 19]
```diff
@@ -116,12 +151,21 @@ def __init__(self, collection: Collection,
...
 
+
 class FindAndModifyCommand(BaseCommand):
     def __init__(self, collection: Collection,
                  kwargs):
-        super().__init__(collection.name)
+        super().__init__(collection.name, kwargs.pop("collation", None))
         for key, value in kwargs.items():
-            self.command_document[key] = value
+            if key == "update" and kwargs.get("replacement", None) is not None:
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:160]: I think you can remove all the "replacement" references here because we never pass "replacement" through kwargs.

File: pymongoexplain/commands.py (hunk 11 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:58]: I refactored it incorrectly, it is done now.

File: pymongoexplain/commands.py (hunk 16 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:130]: Done.

File: pymongoexplain/commands.py (hunk 19 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:160]: Replacement is passed through kwargs for find_one_and_replace

File: pymongoexplain/explainable_collection.py (hunk 18 above)
Comment (juliusgeo) [reply pymongoexplain/explainable_collection.py:149]: Done.

File: pymongoexplain/utils.py (hunk 17 above)
Comment (juliusgeo) [reply pymongoexplain/utils.py:38]: Done.

File: pymongoexplain/commands.py (hunk 19 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:168]: Nevermind, I was confused. Fixed it now.

File: pymongoexplain/commands.py (hunk 11 above)
Comment (ShaneHarvey) [reply pymongoexplain/commands.py:59]: I still don't think this is complete. We can replace `self.command_document["updates"][0]` with `update_doc` now.

File: setup.py [hunk 20]
```diff
@@ -26,8 +26,8 @@
         'console_scripts': [
             'pymongoexplain=pymongoexplain.cli_explain:cli_explain'],
     },
-    tests_require=["pymongo==3.10.1"],
-    install_requires=['pymongo==3.10.1'],
+    tests_require=["pymongo==3.11.0rc0"],
+    install_requires=['pymongo==3.11.0rc0'],
```
Comment (ShaneHarvey) [inline setup.py:30]: I think this should be `pymongo>=3.11`

File: pymongoexplain/commands.py [hunk 21]
```diff
@@ -20,15 +20,20 @@
 
 from bson.son import SON
 from pymongo.collection import Collection
+from pymongo.helpers import _index_document, _fields_list_to_dict
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py:23]: Even though we maintain pymongo we can't rely on any private methods (in python names starting with `_` mean private). Could you vendor (copy) these two methods instead?

Comment (juliusgeo) [reply pymongoexplain/commands.py:23]: Done.

File: pymongoexplain/commands.py (hunk 11 above)
Comment (juliusgeo) [reply pymongoexplain/commands.py:59]: Done.

File: setup.py (hunk 20 above)
Comment (juliusgeo) [reply setup.py:30]: Done.

File: pymongoexplain/commands.py [hunk 22]
```diff
@@ -19,7 +19,6 @@
 from typing import Union
 
 from bson.son import SON
-from bson.py3compat import abc, iteritems, string_type
```
Comment (ShaneHarvey) [inline pymongoexplain/commands.py]: You should replace `from bson.py3compat import abc` with `from collections import abc` instead of using dict/list/set directly.

File: pymongoexplain/commands.py [hunk 23]
```diff
@@ -19,16 +19,71 @@
 from typing import Union
 
 from bson.son import SON
+from bson.py3compat import abc, iteritems, string_type
```
[the remark this answers is not in the document]
Comment (juliusgeo) [reply pymongoexplain/commands.py]: Done.
